src/persist-client/src/impl/gc.rs

```diff
@@ -0,0 +1,215 @@
+        if !should_gc {
+            return;
+        }
+
+        // Spawn GC in a background task, so the state change that triggered it
+        // isn't blocked on it.
+        let gc_span = debug_span!(parent: None, "gc_and_truncate", shard_id=%req.shard_id);
+        gc_span.follows_from(&Span::current());
+
+        let consensus = Arc::clone(&self.consensus);
+        let blob = Arc::clone(&self.blob);
+        let metrics = Arc::clone(&self.metrics);
+        let _ = mz_ore::task::spawn(
+            || "persist::gc_and_truncate",
+            async move { Self::gc_and_truncate(consensus, blob, metrics, req).await }
+                .instrument(gc_span),
+        );
+    }
+
+    pub async fn gc_and_truncate(
+        consensus: Arc<dyn Consensus + Send + Sync>,
+        blob: Arc<dyn Blob + Send + Sync>,
+        metrics: Arc<Metrics>,
+        req: GcReq,
+    ) {
+        // NB: Because these requests can be processed concurrently (and in
+        // arbitrary order), all of the logic below has to work even if we've
+        // already gc'd and truncated past new_seqno_since.
+
+        let deadline = Instant::now() + FOREVER;
+        let path = req.shard_id.to_string();
+        let state_versions = retry_external(&metrics.retries.external.gc_scan, || async {
+            consensus.scan(deadline, &path, SeqNo::minimum()).await
+        })
+        .await;
+
+        debug!(
+            "gc {} for [{},{}) got {} versions from scan",
+            req.shard_id,
+            req.old_seqno_since,
+            req.new_seqno_since,
+            state_versions.len()
+        );
+
+        let mut deleteable_blobs = HashSet::new();
+        for state_version in state_versions {
+            if state_version.seqno < req.new_seqno_since {
+                Self::for_all_keys(&state_version, |key| {
+                    // It's okay (expected) if the key already exists in
+                    // deleteable_blobs, it may have been present in previous
+                    // versions of state.
+                    deleteable_blobs.insert(key.to_owned());
+                });
+            } else if state_version.seqno == req.new_seqno_since {
+                Self::for_all_keys(&state_version, |key| {
+                    // It's okay (expected) if the key doesn't exist in
+                    // deleteable_blobs, it may have been added in this version
+                    // of state.
+                    let _ = deleteable_blobs.remove(key);
+                });
+            } else {
+                // Sanity check the loop logic.
+                assert!(state_version.seqno > req.new_seqno_since);
+                break;
+            }
+        }
+
+        // TODO: Until we do the TODO in State::seqno_since, it's not safe to
+        // actually delete these blobs.
+        deleteable_blobs.clear();
+
+        for key in deleteable_blobs {
+            retry_external(&metrics.retries.external.gc_delete, || async {
```

Comment:
Might be worth it to eventually add a bulk delete call. If the S3 API/the client we use supports this.

Reply:
sorry, not sure I follow! to get a sense for how large a batch is, the metric would need to be a histogram, which turns out to be stupidly expensive (certainly not worth it for this). I'm also not sure it matters how much work each async background task does, so much as what the overall rate of deletions is (which we already track as a counter)